### DISCUSSION TITLE
app.dependency_overrides[get_db] = override_get_db is not working

### DISCUSSION THREAD

**Discussion by pausanchezv** [fastapi/fastapi#8857] [Questions] [answered]
### The problem

Hi, I have followed the tutorials on the fastapi website where I have found everything easy and clear. However, there is one thing I cannot make work and it looks like a lot of people are having this issue. Still, I have tried all the solutions there are on the internet and could not make it work.

I'm trying to run pytest in my fastapi app but even if I follow the tutorial line by line it doesn't work. The fake database gets created but still, the real one is being used by the tests (which is a huge problem). 

The point is if I set a breakpoint inside of the function override_get_db, it never gets there, it is being totally ignored, the rest of the lines are reached by the debugger, so I guess this is the problem. Can anyone who has had the same issue and solved it help, please?

This is my code:

```
import json

from fastapi.testclient import TestClient
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from starlette import status

from database import get_db, Base
from main import app

TEST_DATABASE_URL = "sqlite:///./test.db"
test_engine = create_engine(
   TEST_DATABASE_URL, connect_args={"check_same_thread": False}
)
TestingSessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=test_engine)
Base.metadata.create_all(bind=test_engine)


def override_get_db():
     try:
         db = TestingSessionLocal()
         yield db
     finally:
         db.close()

app.dependency_overrides[get_db] = override_get_db
client = TestClient(app)

def test_get_course():
    response = client.post(
        app.url_path_for('create_course'),
        json={"name": "q1", "level": "q1", "description": "desc", "background_color": "bg"},
    )

    response_data = json.loads(response.content)

    response = client.get(app.url_path_for('get_course', **{"course_id": 1}))
    response_data = json.loads(response.content)
    assert response.status_code == status.HTTP_200_OK
    print(response_data)
    #assert response.json() == {'FastApi': 'UOE PRO'}
```

I'm using Python 3.8.3 and these are the versions I have installed in my virtual environment:

```
fastapi==0.65.1
SQLAlchemy==1.4.15
psycopg2-binary==2.8.6
uvicorn==0.13.4
pydantic==1.8.2
starlette==0.14.2
passlib==1.7.4
bcrypt==3.2.0
python-jose==3.2.0
pytest==6.2.4
requests==2.25.1
```

Thank you so much!

**Comment by teplinsky-maxim** [comment]
Same problem here... just copied the example from the docs, and it does not work. The problem of the docs is that they are so synthetic, no real problem are described more than 'test your hello world made in one file'

**Comment by Ecaloota** [comment]
I'm also having this problem

**Comment by oligond** [comment]
@pausanchezv In order to get help with your issue, you might want to follow some best practices and be precise and helpful to your volunteers with your request. Some hopefully helpful remarks on your post:

You claim to follow the tutorial:
> even if I follow the tutorial line by line it doesn't work

First of all, you might **be more specific about your sources** (these lines you're talking about in the quote), e.g. by providing one or more relevant links to the tutorial pages you're referring to, let's say:
- [Setting up the SQL databases example](https://fastapi.tiangolo.com/tutorial/sql-databases/)
- [Adding a test using `dependency_overrides` ](https://fastapi.tiangolo.com/advanced/testing-database/)

People might be much more eager to help out if you reduce the trouble it takes to figure out what page you're actually on.

Secondly, your claim doesn't seem to be true at all - the sample code you provided has been modified in comparison to the original example:
- you have additional imports such as `from starlette import status`
- some imports appear to be modified: you're importing `get_db` from `.database` whereas it comes from `.main` in the original example (so your version should not run at all if we assume that the other files weren't touched)
- some names are changed
- the test code is changed

Even though most of those changes might be not related to your issue, some might. The obvious modifications you've done to the test file at least provoke questions concerning further changes that you might have done to other files that are part of the tutorial example - we really can't tell! 

As for me, I braindeadly copied the tutorial example, ran pytest (with an empty `sql_app.db`), the test passed, and the record created during the test session was there in `test.db`. In one word: all as expected, couldn't reproduce your issue.

I'd recommend to you to start the same way without your modifications and try again.

Or, alternatively, provide us with some more information about changes you made and the commands you're using to execute your tests (the tutorial does not explicitly provide these commands, so perhaps the mistake can be found there?).

Hope this helps!

BTW: For syntax highlighting, add a language declaration to your code snippet, like
~~~
```python
import this
```
~~~
which will be rendered as

```python
import this
```

**Comment by Ecaloota** [comment]
@oligond, just to verify the above, I went and copied each of the files given at the bottom of the [Setting up the SQL databases example](https://fastapi.tiangolo.com/tutorial/sql-databases/) page, and then created the tests subdirectory according to the format outlined in [Adding a test using dependency_overrides](https://fastapi.tiangolo.com/advanced/testing-database/) - so each file is identical to those given in the tutorial - and I still have the posted issue, wherein the production database is being initialised alongside the test database. For reference, all I did was copy these files verbatim, navigate to /sql_app/ and run pytest, so I haven't attached the code itself.

Is this the intended behaviour? I would have thought that the production database should not be touched, much less initialised, if the dependency_overrides was operating as intended.

In fairness, it appears that only the test database is being written to, but it is a bit frustrating that the production database is being created for each testing session.

**Comment by oligond** [comment, marked as answer]
@Ecaloota Thanks for specifying the issue, I had really misunderstood the original post.

Now I get what your question _actually_ is about 😉 

The "production/real" database is created when running the test just because of the simple/naive setup of the example. 

**So what's going on?**

- the module containing the test code is importing stuff from `..main`
- this means that all top level code in `..main` is executed when running the test code
- the top level code in `..main` -- among other things ­-- creates the database as demonstrated here:
   <https://fastapi.tiangolo.com/tutorial/sql-databases/#create-the-database-tables>

**How to fix that?**

Prevent the database creation code from being executed on module import!

As a quick and dirty way to achieve this, let's wrap it into a function. We're only interested in getting the test to run as we expect, so we don't have to care too much. You could as well just comment it out, if all you want is that the example behaves well for your test.

A real life app wouldn't usually just create its database on startup, a more common way would be to define some task (together with other tasks relating to database housekeeping) that you run manually or programmatically as required.

So that's our modified `main.py`:

```python
from typing import List

from fastapi import Depends, FastAPI, HTTPException
from sqlalchemy.orm import Session

from . import crud, models, schemas
from .database import engine, SessionLocal


app = FastAPI()


# --- begin modified: won't be executed on module import ---
def create_db():
    models.Base.metadata.create_all(bind=engine)
# --- end modified ---


# Dependency
def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


@app.post("/users/", response_model=schemas.User)
def create_user(user: schemas.UserCreate, db: Session = Depends(get_db)):
    db_user = crud.get_user_by_email(db, email=user.email)
    if db_user:
        raise HTTPException(status_code=400, detail="Email already registered")
    return crud.create_user(db=db, user=user)


@app.get("/users/", response_model=List[schemas.User])
def read_users(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    users = crud.get_users(db, skip=skip, limit=limit)
    return users


@app.get("/users/{user_id}", response_model=schemas.User)
def read_user(user_id: int, db: Session = Depends(get_db)):
    db_user = crud.get_user(db, user_id=user_id)
    if db_user is None:
        raise HTTPException(status_code=404, detail="User not found")
    return db_user


@app.post("/users/{user_id}/items/", response_model=schemas.Item)
def create_item_for_user(
    user_id: int, item: schemas.ItemCreate, db: Session = Depends(get_db)
):
    return crud.create_user_item(db=db, item=item, user_id=user_id)


@app.get("/items/", response_model=List[schemas.Item])
def read_items(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    items = crud.get_items(db, skip=skip, limit=limit)
    return items

```
With the modification in place, delete the `sql_app.db` and run pytest again. The file shouldn't be recreated anymore.

**Comment by Ecaloota** [comment]
@oligond Thank you for the above, this all makes sense and I've gotten it to work. Thanks again!

**Comment by pausanchezv** [comment]
> @pausanchezv In order to get help with your issue, you might want to follow some best practices and be precise and helpful to your volunteers with your request. Some hopefully helpful remarks on your post:
> 
> You claim to follow the tutorial:
> 
> > even if I follow the tutorial line by line it doesn't work
> 
> First of all, you might **be more specific about your sources** (these lines you're talking about in the quote), e.g. by providing one or more relevant links to the tutorial pages you're referring to, let's say:
> 
> * [Setting up the SQL databases example](https://fastapi.tiangolo.com/tutorial/sql-databases/)
> * [Adding a test using `dependency_overrides` ](https://fastapi.tiangolo.com/advanced/testing-database/)
> 
> People might be much more eager to help out if you reduce the trouble it takes to figure out what page you're actually on.
> 
> Secondly, your claim doesn't seem to be true at all - the sample code you provided has been modified in comparison to the original example:
> 
> * you have additional imports such as `from starlette import status`
> * some imports appear to be modified: you're importing `get_db` from `.database` whereas it comes from `.main` in the original example (so your version should not run at all if we assume that the other files weren't touched)
> * some names are changed
> * the test code is changed
> 
> Even though most of those changes might be not related to your issue, some might. The obvious modifications you've done to the test file at least provoke questions concerning further changes that you might have done to other files that are part of the tutorial example - we really can't tell!
> 
> As for me, I braindeadly copied the tutorial example, ran pytest (with an empty `sql_app.db`), the test passed, and the record created during the test session was there in `test.db`. In one word: all as expected, couldn't reproduce your issue.
> 
> I'd recommend to you to start the same way without your modifications and try again.
> 
> Or, alternatively, provide us with some more information about changes you made and the commands you're using to execute your tests (the tutorial does not explicitly provide these commands, so perhaps the mistake can be found there?).
> 
> Hope this helps!
> 
> BTW: For syntax highlighting, add a language declaration to your code snippet, like
> 
> ```
> ```python
> import this
> ```
> ```
> 
> which will be rendered as
> 
> ```python
> import this
> ```

I found a workaroud like three months ago, but I'll take into account how to comment an issue in the future. Thanks.
